ports/stm32f4/common-hal/microcontroller/__init__.c

```diff
@@ -39,17 +39,56 @@
 #include "supervisor/filesystem.h"
 #include "supervisor/shared/safe_mode.h"
 
-// This routine should work even when interrupts are disabled. Used by OneWire
-// for precise timing.
+#include "stm32f4xx_hal.h"
+
+//tested divisor value for busy loop in us delay
+#define LOOP_TICKS 12
+
+STATIC uint32_t get_us(void) {
+    uint32_t ticks_per_us = HAL_RCC_GetSysClockFreq()/1000000;
+    uint32_t micros, sys_cycles;
+    do {
+        micros = ticks_ms;
+        sys_cycles = SysTick->VAL; //counts backwards
+    } while (micros != ticks_ms); //try again if ticks_ms rolled over
+    return (micros * 1000) + (ticks_per_us * 1000 - sys_cycles) / ticks_per_us;
+}
+
 void common_hal_mcu_delay_us(uint32_t delay) {
-  //TODO: implement equivalent of mp_hal_delay_us(delay);
-  //this is fairly annoying in the STM32 HAL
+    if (__get_PRIMASK() == 0x00000000) {
+        //by default use ticks_ms
+        uint32_t start = get_us();
+        while (get_us()-start < delay) {
+            __asm__ __volatile__("nop");
+        }
+    } else {
+        //when SysTick is disabled, approximate with busy loop
+        const uint32_t ucount = HAL_RCC_GetSysClockFreq() / 1000000 * delay / LOOP_TICKS;
+        for (uint32_t count = 0; ++count <= ucount;) {
+        }
+    }
 }
```

Comment:
In `atmel-samd/mphalport.c`, there's a simpler way that just has a NOP loop. It looks like `get_us()` could be relatively expensive. What made you choose this way over the atmel-samd way, or over the `CYCCNT` way that seems to be typical in the STM32 examples I've seen in websearches?

Reply:
@dhalbert atmels mphalport.c assumes a set system clock per processor which is not currently guaranteed among the STM32 boards we support. But it's essentially the same thing as my implementation for when interrupts are disabled. 

The approach overall is based on micropython, stm32duino, and the recommendations I saw on ST forums - I didn't personally find anyone recommending the DWT for this purpose off google. It also has the advantage of never overflowing, which the DWT will eventually do if you plug in irresponsible numbers. However, just researching it now, it does look like stm32duino switched to the DWT this [march](https://github.com/stm32duino/Arduino_Core_STM32/pull/477). It seems reason they and MPy didn't have it by default was the fact that M0/M0+ MCUs don't have it, and possibly avoiding the setup process. 

If the DWT is already used or initialized anywhere in CPy I'd need to know about that. Otherwise I can either set it up as a static instance here in Microcontroller, or make a new module for it? What do you think?

Reply:
We had an audio conversation; I'm good with this!